### PR TITLE
New plugin for creating GitOps-managed Kubernetes right from Backstage.

### PR DESCRIPTION
**Pull request by chanwit**
## Plugin - `gitops-profiles` 

This plugin is for creating and managing GitOps-managed Kubernetes cluster. It works by creating a GitOps repository on GitHub then provision new (for now EKS) cluster from the cluster config.

#### :heavy_check_mark: Checklist
- [x] All tests are passing `yarn test`
- [x] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality

Also with a document on how to deploy this plugin to Kubernetes:
https://github.com/spotify/backstage/blob/74b6a62221e56dde28dd711ef5f6375dacf5bb22/plugins/gitops-profiles/README.md

### CODE REVIEW

**Comment by chanwit** [comment]
Plugin added to the Explorer page.

**Comment by chanwit** [comment]
Listing GitOps-based clusters from GitHub.

Listing GitOps-based clusters from GitHub

**Comment by chanwit** [comment]
Cluster creation step 1, select Cluster Template (from a GitHub repo).

Cluster Creation Step 1

**Comment by chanwit** [comment]
Cluster creation step 2, select ready-to-use profiles to apply.

Cluster creation step 2

**Comment by chanwit** [comment]
Cluster creation step 3, fill target repository and keys.

Cluster creation step 3

**Comment by chanwit** [comment]
Then wait for GitHub Actions to kick in and provision a new cluster for us.

Wait

**Comment by chanwit** [comment]
Here's a GitOps repository and the cluster created by this plugin.

https://github.com/chanwit/new-cluster

**Comment by chanwit** [comment]
/cc @leemills83 @soapraj @Rugvip 
Could you ptal?

File: plugins/gitops-profiles/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+This method of serving the plugin provides quicker iteration speed and a faster startup and hot reloads.
+It is only meant for local development, and the setup for it can be found inside the [/dev](/dev) directory.
+
+## Deployment
+
+Please follow these steps to build Docker images for deployment.
+
+At Backstage directory,
+
+- Replace Backstage's NGINX config with this plugin's config
```
Comment (Rugvip) [inline plugins/gitops-profiles/README.md:22]: Please make sure the backend can be called directly for now, as in setting CORS headers + config on the frontend. Proxying will be part of the backend but isn't there yet.

File: plugins/gitops-profiles/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  $ cp plugins/gitops-profiles/deploy/default.conf.template docker/default.conf.template
+```
+
+- Create Backstage images with
+
+```bash
+  $ yarn docker-build
+  $ export REPO=<your_docker_id>
+  $ docker tag spotify/backstage $REPO/gitops-app-backstage
+  $ docker push $REPO/gitops-app-backstage
```
Comment (Rugvip) [inline plugins/gitops-profiles/README.md:34]: This would do better in top-level configuration as it's not plugin-specific.

**Review by Rugvip — COMMENTED** [review]
Few questions before diving into the code

File: plugins/gitops-profiles/deploy/gitops-app-manifest.yaml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  selector:
+    matchLabels:
+      app: gitops-api
+  template:
+    metadata:
+      labels:
+        app: gitops-api
+    spec:
+      containers:
+      - image: chanwit/gitops-api:latest
```
[the remark this answers is not in the document]
Comment (chanwit) [reply plugins/gitops-profiles/deploy/gitops-app-manifest.yaml:41]: It's the backend for GitOps APIs written in Golang. But I tweaked the build system to use `yarn` just like the `esbuild` did.

Here's the repo: https://github.com/chanwit/gitops-api

I'll be happy if this Golang project will be included into the Backstage tree too (to show the diversity of backends supported by Backstage). Happy to discuss about how could we include this kind of plugin backend into the project.

**Comment by chanwit** [comment]
#### :heavy_check_mark: Checklist
- [x] Remove nginx config, instead allow the backend URL to be configurable, similar to the lighthouse plugin
- [x] Make sure CORS is enabled in the backend so that the plugin does not need to talk through a proxy
- [x] The backend implementation and packaging can be left as is, but add some lightweight documentation
- [x] Update the plugin docs to be from the point of view of running the plugin as a part of a backstage deployment
- [x] Move the k8s manifest for the backend to the readme and reference it as an example of how to deploy it in k8s. We don’t need to include the other K8s manifests.

/cc @leemills83 @Rugvip

File: plugins/gitops-profiles/README.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Your plugin has been added to the example app in this repository, meaning you'll be able to access it by running `yarn start` in the root directory, and then navigating to [/gitops-clusters](http://localhost:3000/gitops-profiles).
+
+You can also serve the plugin in isolation by running `yarn start` in the plugin directory.
+This method of serving the plugin provides quicker iteration speed and a faster startup and hot reloads.
+It is only meant for local development, and the setup for it can be found inside the [/dev](/dev) directory.
+
+## Use GitOps-API backend with Backstage
+
+The backend of this plugin is written in Goland and its source code is available [here](https://github.com/chanwit/gitops-api) as a separate GitHub repository.
```
Comment (Rugvip) [inline plugins/gitops-profiles/README.md:18]: Go/Golang*, I don't care what IDE you're using :grin:

File: plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+          setShowProgress(false);
+        }
+      }, 10000);
+      return () => clearInterval(interval);
+    }
+    return () => {};
+  }, [pollingLog]);
+
+  if (params.owner === undefined || params.repo === undefined) {
+    const { loading, error, value } = useAsync<ListClusterStatusesResponse>(
```
Comment (Rugvip) [inline plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx:80]: Can't call this hook conditionally

File: plugins/gitops-profiles/src/components/ProfileCatalog/ProfileCatalog.tsx [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+          message: msg,
+          conclusion: 'success',
+        },
+      ]),
+    );
+  };
+
+  const doCreateCluster = async () => {
+    // eslint-disable-next-line no-console
+    console.log(gitopsProfiles);
```
Comment (Rugvip) [inline plugins/gitops-profiles/src/components/ProfileCatalog/ProfileCatalog.tsx:181]: Kill it with 🔥

File: plugins/gitops-profiles/README.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+The binary of this plugin is available as a ready-to-use Docker image, [https://hub.docker.com/chanwit/gitops-api](https://hub.docker.com/chanwit/gitops-api).
+To start using GitOps with Backstage, you have to start the backend using the following command:
+
+```bash
+$ docker run -d --init -p 3008:8080 chanwit/gitops-api
+```
+
+Please note that this plugin requires the backend to run on port 3008.
+
+## Deploy on Kubernetes
```
Comment (Rugvip) [inline plugins/gitops-profiles/README.md:28]: Actually intended for this to be part of the backend docs instead, if it's not too much trouble :grin:

**Review by Rugvip — COMMENTED** [review]
👍, few minor comments

File: plugins/gitops-profiles/README.md (hunk 4 above)
Comment (chanwit) [reply plugins/gitops-profiles/README.md:18]: lol, sorry

File: plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx (hunk 5 above)
Comment (chanwit) [reply plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx:80]: could you collaborate this?

File: plugins/gitops-profiles/README.md (hunk 7 above)
Comment (chanwit) [reply plugins/gitops-profiles/README.md:28]: No problem. I'll move it to the backend :-)

File: plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx (hunk 5 above)
Comment (Rugvip) [reply plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx:80]: You shouldn't call hooks in conditional code, see plugins/auth-backend/src/providers/OAuthProvider.test.ts

Comment (chanwit) [reply plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx:80]: acked

Comment (chanwit) [reply plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx]: fixed and I had to refactor ClusterPage component into two to avoid the if.

Comment (chanwit) [reply plugins/gitops-profiles/src/components/ClusterPage/ClusterPage.tsx]: thank you for the comment, @Rugvip!

**Comment by chanwit** [comment]
rebased to master and put fixes to make lint happy.

**Comment by swiftdiaries** [comment]
Hey @chanwit,
Quick question, is this specific to EKS + FireKube?

**Comment by chanwit** [comment]
@swiftdiaries the GitOps provisioner is now implemented as GitHub Action to start EKS clusters.
No Firekube inside atm.

**Comment by swiftdiaries** [comment]
Thank you 😊

**Review by Rugvip — APPROVED** [review]
👍, @soapraj, wanna have a look as well?

**Comment by chanwit** [comment]
Happy to wait for @soapraj comments too :-) :-)

**Review by soapraj — APPROVED** [review]

**Comment by soapraj** [comment]
LGTM 🚀

**Pull request merged**